lib/shopify_api/rest.ex

```diff
@@ -9,27 +9,68 @@ defmodule ShopifyAPI.REST do
 
   alias ShopifyAPI.AuthToken
   alias ShopifyAPI.JSONSerializer
+  alias ShopifyAPI.REST.PaginatedRequest
   alias ShopifyAPI.REST.Request
 
-  @doc false
-  def get(%AuthToken{} = auth, path, params \\ []) do
-    Request.perform(auth, :get, path, "", params)
+  @doc """
+  Underlying utility retrieval function. The options passed affect both the
+  return value and, ultimately, the number of requests made to Shopify.
+  Options:
+    * `:pagination` - Can be `:none`, `:stream`, or `:auto`. Defaults to :auto
+      `:none` will only return the first page. You won't have access to the headers to manually
+      paginate.
+      `:auto` will block until all the pages have been retrieved and concatenated together.
+      `:stream` will return a `Stream`, prepopulated with the first page.
+  """
+  def get(%AuthToken{} = auth, path, params \\ [], options \\ []) do
+    case pagination(options) do
+      :none ->
+        with {:ok, response} <- Request.perform(auth, :get, path, "", params) do
+          {:ok, fetch_body(response)}
+        end
+
+      :stream ->
+        PaginatedRequest.product_stream(auth.shop_name, auth.token)
+
+      _auto_or_nil ->
+        Enum.to_list(PaginatedRequest.product_stream(auth.shop_name, auth.token))
+    end
   end
 
   @doc false
   def post(%AuthToken{} = auth, path, object \\ %{}) do
     with {:ok, body} <- JSONSerializer.encode(object),
-         do: Request.perform(auth, :post, path, body)
+         {:ok, response} <- Request.perform(auth, :post, path, body),
+         response_body <- fetch_body(response) do
+      {:ok, response_body}
+    end
   end
 
   @doc false
   def put(%AuthToken{} = auth, path, object) do
     with {:ok, body} <- JSONSerializer.encode(object),
-         do: Request.perform(auth, :put, path, body)
+         {:ok, response} <- Request.perform(auth, :put, path, body),
+         response_body <- fetch_body(response) do
+      {:ok, response_body}
+    end
   end
 
   @doc false
   def delete(%AuthToken{} = auth, path) do
-    Request.perform(auth, :delete, path)
+    with {:ok, response} <- Request.perform(auth, :delete, path),
```

Comment:
These all look identical, could be an opportunity to condense them

Reply:
@hez I'm not sure what you mean by this, do you have a suggestion for how condensing them could work?